File: DeepSpace-Ruby/lib/SpaceStation.rb
# encoding:utf-8
require_relative "SuppliesPackage"
require_relative "Hangar"
require_relative "Damage"
require_relative "Weapon"
require_relative "WeaponType"
require_relative "Loot"
require_relative "ShieldBooster"
require_relative "CardDealer"
require_relative "SpaceStationToUI"


module Deepspace
class SpaceStation
   @@MAXFUEL = 100
   @@SHIELDLOSSPERUNITSHOT = 0.1

#
# ────────────────────────────────────────────────────────────── CONSTRUCTOR ─────
#

   def initialize (n, supplies)
      @name        = n
      @nMedals     = 0
      @ammoPower   = supplies.ammoPower
      @fuelUnits   = assignFuelValue(supplies.fuelUnits)
      @shieldPower = supplies.shieldPower

      @hangar        = nil
      @pendingDamage = nil

      @weapons        = Array.new
      @shieldBoosters = Array.new
   end

     def createStation(station)
      @name           = station.name
      @ammoPower      = station.ammoPower
      @nMedals        = station.nMedals
      @shieldPower    = station.shieldPower
      @pendingDamage  = station.pendingDamage
      @weapons        = station.weapons
      @shieldBoosters = station.shieldBoosters
      @hangar         = station.hangar
      assignFuelValue(station.fuelUnits)
    end

#
# ───────────────────────────────────────────────────────── METODOS PRIVADOS ─────
#

   private def assignFuelValue (f)
      @fuelUnits = [@@MAXFUEL, f].min
   end

   private def cleanPendingDamage()
      if (@pendingDamage != nil && @pendingDamage.hasNoEffect)
         @pendingDamage = nil
      end
   end

#
# ───────────────────────────────────────────────────────────────── DISCARDS ─────
#


   def discardHangar()
         @hangar = nil
   end

   def discardWeapon (i)
        size = @weapons.size
        #indice = 0
        if i >= 0 and i < size
            w = @weapons.delete_at(i)
            if (@pendingDamage != nil)
                @pendingDamage.discardWeapon(w)
                cleanPendingDamage
            end
        end
   end

   def discardWeaponInHangar(i)
      if (@hangar != nil && i >= 0)
         @hangar.removeWeapon(i)
      end
   end

   def discardShieldBooster (i)
        size = @shieldBoosters.size
        #indice = 0
        if i >= 0 and i < size
            sh = @shieldBoosters.delete_at(i)
            if (@pendingDamage != nil)
                @pendingDamage.discardShieldBooster
                cleanPendingDamage
            end
        end
   end


   def discardShieldBoosterInHangar(i)
      if(@hangar != nil && i >= 0)
         @hangar.removeShieldBooster(i)
      end
   end

#
# ─────────────────────────────────────────────────────────────── INTERFACES ─────
#

   attr_reader :nMedals, :name, :fuelUnits, :ammoPower, :hangar, :shieldBoosters, :weapons, :shieldPower, :pendingDamage

   def getSpeed
      @fuelUnits/@@MAXFUEL
   end

   def getUIversion
      SpaceStationToUI.new(self)
   end


#
# ─────────────────────────────────────────────────────────────── RECEPTORES ─────
#

   def receiveWeapon(w)
      if @hangar != nil
         @hangar.addWeapon(w)
      else
         false
      end
   end

   def receiveShieldBooster(s)
      if (@hangar != nil)
         @hangar.addShieldBooster(s)
         true
      else
         false
      end
   end

   def receiveHangar(h)
      if (@hangar == nil)
         @hangar = h
      end
   end

   def receiveSupplies(s)
      @ammoPower   += s.ammoPower
      @shieldPower += s.shieldPower
      assignFuelValue(s.fuelUnits + @fuelUnits)
   end

   def receiveShot (shot)
        myprotection = protection
        @shieldPower -= @@SHIELDLOSSPERUNITSHOT*shot
        @shieldPower = [@shieldPower, 0.0].max
        if (shot <= myprotection)
            ShotResult::RESIST
        else
            ShotResult::DONOTRESIST
        end
   end

#
# ────────────────────────────────────────────────────────────────── SETTERS ─────
#

   def setPendingDamage(d)
      @pendingDamage = d.adjust(@weapons,@shieldBoosters)
   end

   def setLoot (s)
        dealer = CardDealer.instance

        h = s.nHangars
        @hangar = dealer.nextHangar
        if (h > 0)
            receiveHangar(h)
        end

        indice = 0
        elements = s.nSupplies
        while indice < elements
           sup = dealer.nextSuppliesPackage
           receiveSupplies(sup)
           indice += 1
        end

        indice = 0
        elements = s.nWeapons
        while indice < elements
           w = dealer.nextWeapon
           receiveWeapon(w)
           indice += 1
        end

        indice = 0
        elements = s.nShields

        while indice < elements
            sh = dealer.nextShieldBooster
            receiveShieldBooster(sh)
           indice += 1
        end

        @nMedals += s.nMedals

        if s.efficient
            Transformation::GETEFFICIENT
        elsif s.spaceCity
            Transformation::SPACECITY
        else
            Transformation::NOTRANSFORM
        end
   end

#
# ─────────────────────────────────────────────────────────────────── MOUNTS ─────
#

   def mountWeapon(i)
      if (i >= 0)
         if (@hangar != nil)
            arma = @hangar.removeWeapon(i)

            if (arma != nil)
               @weapons.push(arma)
            end
         end
      end
   end

   def mountShieldBooster(i)
      if (i >= 0)
         if (@hangar != nil)
            escudo = @hangar.removeShieldBooster(i)

            if (escudo != nil)
               @shieldBoosters.push(escudo)
            end
         end
      end
   end

#
# ──────────────────────────────────────────────────────────────────── OTROS ─────
#

   def move
      @fuelUnits -= getSpeed
   end

   def fire
        factor = 1
        for w in @weapons
            factor *=w.useIt
        end

        @ammoPower*factor
   end

   def protection
        factor = 1
        for s in @shieldBoosters
            factor *=s.useIt
        end

        @shieldPower*factor
   end

   def validState
      @pendingDamage == nil or @pendingDamage.hasNoEffect
   end

   def cleanUpMountedItems
      for arma in @weapons
         if(arma.uses == 0)
            @weapons.delete(arma)
          end
      end
      for escudo in @shieldBoosters
         if(escudo.uses == 0)
            @shieldBoosters.delete(escudo)
         end
      end
   end

   def to_s
        getUIversion.to_s
   end
end
end

File: DeepSpace-Java/src/deepspace/SpaceStation.java
package deepspace;

import java.util.ArrayList;

// import sun.security.jca.GetInstance;

class SpaceStation implements SpaceFighter{
    private static int MAXFUEL = 100;
    private static float SHIELDOSSPERUNITSHOT = 0.1f;

    private int nMedals;
    private String name;
    private float fuelUnits;
    private float ammoPower;
    private float shieldPower;

    private Hangar hangar;
    private Damage pendingDamage = new NumericDamage(0, 0);
    private ArrayList<Weapon> weapons = new ArrayList<>();
    private ArrayList<ShieldBooster> shieldBoosters = new ArrayList<>();


//
// ───────────────────────────────────────────────────────── METODOS PRIVADOS ─────
//

    private void assignFuelValue (float f) {
        fuelUnits = Math.min(f, MAXFUEL);
    }

    private void cleanPendingDamage () {
        if (pendingDamage.hasNoEffect())
            pendingDamage = null;
    }

//
// ────────────────────────────────────────────────────────────── CONSTRUCTOR ─────
//

    SpaceStation (String n, SuppliesPackage supplies) {
        name        = n;
        ammoPower   = supplies.getAmmoPower();
        fuelUnits   = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
    }

    SpaceStation (SpaceStation estacion) {
        name           = estacion.name;
        ammoPower      = estacion.ammoPower;
        fuelUnits      = estacion.fuelUnits;
        shieldPower    = estacion.shieldPower;
        nMedals        = estacion.nMedals;
        pendingDamage  = estacion.pendingDamage;
        weapons        = estacion.weapons;
        shieldBoosters = estacion.shieldBoosters;
        hangar         = estacion.hangar;
    }

//
// ───────────────────────────────────────────────────────────────── DISCARDS ─────
//

    public void discardShieldBooster (int i) {
        if (i >= 0 && i < shieldBoosters.size()) {
            shieldBoosters.remove(i);

            if (pendingDamage != null) {
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }
        }
    }

    public void discardShieldBoosterInHangar (int i) {
        if (hangar != null)
            hangar.removeShieldBooster(i);
    }

    public void discardWeapon (int i) {
        if (i >= 0 && i < weapons.size()) {
            Weapon w = weapons.remove(i);

            if (pendingDamage != null) {
                pendingDamage.discardWeapon(w); // FIXME
                cleanPendingDamage();
            }
        }
    }

    public void discardWeaponInHangar (int i) {
        if (hangar != null)
            hangar.removeWeapon(i);
    }

    public void discardHangar () {
        hangar = null;
    }

//
// ─────────────────────────────────────────────────────────────── INTERFACES ─────
//

    public float getAmmoPower() {
        return ammoPower;
    }

    public float getFuelUnits() {
        return fuelUnits;
    }

    public Hangar getHangar () {
        return hangar;
    }

    public String getName () {
        return name;
    }

    public int getNMedals() {
        return nMedals;
    }

    public ArrayList<ShieldBooster> getShieldBoosters() {
        return shieldBoosters;
    }

    public ArrayList<Weapon> getWeapons () {
        return weapons;
    }

    public float getShieldPower () {
        return shieldPower;
    }

    public float getSpeed() {
        return fuelUnits/MAXFUEL;
    }

    public Damage getPendingDamage () {
        return pendingDamage;
    }

    public SpaceStationToUI getUIversion () {
        return new SpaceStationToUI(this);
    }


//
// ─────────────────────────────────────────────────────────────── RECEPTORES ─────
//

    public void receiveHangar (Hangar h) {
        if (hangar == null)
            hangar = h;
    }

    public boolean receiveShieldBooster (ShieldBooster s) {
        if (hangar != null)
            return hangar.addShieldBooster(s);

        return false;
    }

    public ShotResult receiveShot (float shot) {
        float myProtection = protection();

        if (myProtection >= shot) {
            shieldPower -= SHIELDOSSPERUNITSHOT * shot;
            shieldPower = Math.max(0, shieldPower);

            return ShotResult.RESIST;
        }
        else {
            shieldPower = 0;

            return ShotResult.DONOTRESIST;
        }
    }

    public void receiveSupplies (SuppliesPackage s) {
        ammoPower   += s.getAmmoPower();
        shieldPower += s.getShieldPower();

        assignFuelValue(fuelUnits + s.getFuelUnits());
    }

    public boolean receiveWeapon (Weapon w) {
        if (hangar != null)
            return hangar.addWeapon(w);

        return false;
    }

//
// ────────────────────────────────────────────────────────────────── SETTERS ─────
//

    public Transformation setLoot (Loot l) {
        CardDealer dealer = CardDealer.getInstance();

        int h = l.getNHangars();

        if (h > 0) {
            Hangar hangar = dealer.nextHangar();
            receiveHangar(hangar);
        }

        int elements = l.getNSupplies();

        for (int i = 0; i < elements; i++) {
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            receiveSupplies(sup);
        }

        elements = l.getNWeapons();

        for (int i = 0; i < elements; i++) {
            Weapon weap = dealer.nextWeapon();
            receiveWeapon(weap);
        }

        elements = l.getNShields();

        for (int i = 0; i < elements; i++) {
            ShieldBooster sh = dealer.nextShieldBooster();
            receiveShieldBooster(sh);
        }

        nMedals += l.getNMedals();


        if ( l.getEfficient() ) {
            return Transformation.GETEFFICIENT;
        }
        else if (l.spaceCity()) {
            return Transformation.SPACECITY;
        }
        else {
            return Transformation.NOTRANSFORM;
        }
    }

    public void setPendingDamage (Damage d) {
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }

    public void mountShieldBooster (int i) {
        if (i >= 0) {
            if (hangar != null) {
                ShieldBooster aMontar = hangar.removeShieldBooster(i);

                if (aMontar != null)
                    shieldBoosters.add(aMontar);
            }
        }
    }

    public void mountWeapon (int i) {
        if (i >= 0 ) {
            if (hangar != null) {
                Weapon aMontar = hangar.removeWeapon(i);

                if (aMontar != null)
                    weapons.add(aMontar);
            }
        }
    }

//
// ──────────────────────────────────────────────────────────────────── OTROS ─────
//

    public boolean validState () {
        return ( pendingDamage == null || pendingDamage.hasNoEffect() );
    }

    public void move () {
        if (fuelUnits > 0)
            fuelUnits -= getSpeed();
    }

    public float protection () {
        int factor = 1;

        for (ShieldBooster s : shieldBoosters) {
            factor *= s.useIt();
        }

        return shieldPower * factor;
    }

    public float fire () {
        int factor = 1;

        for (Weapon w : weapons) {
            factor *= w.useIt();
        }

        return ammoPower * factor;
    }

    public void cleanUpMountedItems () {
        for (Weapon arma: weapons)
            if (arma.getUses() == 0)
                weapons.remove(arma);

        for (ShieldBooster escudo: shieldBoosters)
            if (escudo.getUses() == 0)
                shieldBoosters.remove(escudo);
    }

    @Override
    public String toString() {
        return    "-> Nombre: " + name
                + "\n\t-> Potencia: " + ammoPower
                + "\n\t-> Escudos: " + shieldPower
                + "\n\t-> nMedals: " + nMedals
                + "\n\t-> Daño pendiente:" + pendingDamage.toString()
                + "\n\t-> Gasolina: " + fuelUnits
                + "\n\t-> Hangar: " + hangar.toString();
    }
}